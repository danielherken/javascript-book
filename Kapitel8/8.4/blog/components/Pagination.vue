<template>
  <div class="row">
    <div class="col text-center">
      <b-button :disabled="page === 0" @click="$emit('pageChanged', page - 1)">
        <<
      </b-button>
      <b-button :disabled="((page + 1) * itemsPerPage) >= itemsCount" @click="click">
        >>
      </b-button>
    </div>
  </div>
</template>

<script>
export default {
  name: 'Pagination',
  props: {
    page: {
      type: Number,
      required: true
    },
    itemsCount: {
      type: Number,
      required: true
    },
    itemsPerPage: {
      type: Number,
      required: true
    }
  },
  methods: {
    click() {
      console.log('click');
      this.$emit('pageChanged', this.page + 1);
    }
  }
}
</script>

<style scoped>

</style>
