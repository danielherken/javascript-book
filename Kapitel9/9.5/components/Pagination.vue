<template>
  <div class="row">
    <div class="col text-center">
      <b-button
        variant="primary"
        :disabled="page === 0"
        @click="$emit('pageChanged', page - 1)"
      >
        {{ $t('previous') }}
      </b-button>
      <b-button
        variant="primary"
        :disabled="(page + 1) * itemsPerPage >= itemsCount"
        @click="$emit('pageChanged', page + 1)"
      >
        {{ $t('next') }}
      </b-button>
    </div>
  </div>
</template>

<script>
export default {
  name: 'Pagination',
  props: {
    page: {
      type: Number,
      required: true,
    },
    itemsCount: {
      type: Number,
      required: true,
    },
    itemsPerPage: {
      type: Number,
      required: true,
    },
  },
};
</script>

<style scoped></style>
