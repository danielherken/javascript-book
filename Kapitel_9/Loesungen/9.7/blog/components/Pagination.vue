<template>
  <div class="row">
    <div class="col text-center">
      <b-button
        variant="primary"
        :disabled="page === 0"
        @click="$emit('pageChanged', page - 1)"
      >
        {{ $t('previous') }}
      </b-button>
      <b-button
        variant="primary"
        :disabled="(page + 1) * itemsPerPage >= itemsCount"
        @click="$emit('pageChanged', page + 1)"
      >
        {{ $t('next') }}
      </b-button>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';

@Component
export default class Pagination extends Vue {
  @Prop({ required: true })
  private page!: number;

  @Prop({ required: true })
  private itemsCount!: number;

  @Prop({ required: true })
  private itemsPerPage!: number;
}
</script>

<style scoped></style>
