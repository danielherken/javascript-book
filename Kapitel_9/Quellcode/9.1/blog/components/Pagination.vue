<template>
  <div class="row">
    <div class="col text-center">
      <b-button :disabled="page === 0" @click="$emit('pageChanged', page - 1)">
        Previous
      </b-button>
      <b-button
        :disabled="(page + 1) * itemsPerPage >= itemsCount"
        @click="$emit('pageChanged', page + 1)"
      >
        Next
      </b-button>
    </div>
  </div>
</template>

<script>
export default {
  name: 'Pagination',
  props: {
    page: {
      type: Number,
      required: true,
    },
    itemsCount: {
      type: Number,
      required: true,
    },
    itemsPerPage: {
      type: Number,
      required: true,
    },
  },
};
</script>

<style scoped></style>
