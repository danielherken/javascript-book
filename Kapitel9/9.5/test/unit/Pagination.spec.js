import { mount, createLocalVue } from '@vue/test-utils';
import { BootstrapVue } from 'bootstrap-vue';
import Pagination from '~/components/Pagination.vue';

let localVue;

beforeEach(() => {
  localVue = createLocalVue();
  localVue.use(BootstrapVue);
});

describe('Pagination', () => {
  test('is handling first page correctly', () => {
    const wrapper = mount(Pagination, {
      mocks: {
        $t: (key) => key,
      },
      propsData: {
        page: 0,
        itemsCount: 30,
        itemsPerPage: 10,
      },
      localVue,
    });

    // Tested of die Komponente erstellt werden konnte
    expect(wrapper.vm).toBeTruthy();

    // Prüft ob die Komponente zwei Buttons enthält
    const buttons = wrapper.findAll('button');
    expect(buttons.length).toBe(2);

    // Prüft ob der erste Button deaktiviert ist
    expect(buttons.at(0).element.disabled).toBe(true);

    // Prüft ob der zweite Button aktiviert ist
    expect(buttons.at(1).element.disabled).toBe(false);

    expect(wrapper.element).toMatchSnapshot();
  });

  test('is handling middle page correctly', () => {
    const wrapper = mount(Pagination, {
      mocks: {
        $t: (key) => key,
      },
      propsData: {
        page: 1,
        itemsCount: 30,
        itemsPerPage: 10,
      },
      localVue,
    });

    // Tested of die Komponente erstellt werden konnte
    expect(wrapper.vm).toBeTruthy();

    // Prüft ob die Komponente zwei Buttons enthält
    const buttons = wrapper.findAll('button');
    expect(buttons.length).toBe(2);

    // Prüft ob der erste Button aktiviert ist
    expect(buttons.at(0).element.disabled).toBe(false);

    // Prüft ob der zweite Button aktiviert ist
    expect(buttons.at(1).element.disabled).toBe(false);

    expect(wrapper.element).toMatchSnapshot();
  });

  test('is handling last page correctly', () => {
    const wrapper = mount(Pagination, {
      mocks: {
        $t: (key) => key,
      },
      propsData: {
        page: 2,
        itemsCount: 30,
        itemsPerPage: 10,
      },
      localVue,
    });

    // Tested of die Komponente erstellt werden konnte
    expect(wrapper.vm).toBeTruthy();

    // Prüft ob die Komponente zwei Buttons enthält
    const buttons = wrapper.findAll('button');
    expect(buttons.length).toBe(2);

    // Prüft ob der erste Button aktiviert ist
    expect(buttons.at(0).element.disabled).toBe(false);

    // Prüft ob der zweite Button deaktiviert ist
    expect(buttons.at(1).element.disabled).toBe(true);

    expect(wrapper.element).toMatchSnapshot();
  });

  test('is emitting page changed event', async () => {
    const wrapper = mount(Pagination, {
      mocks: {
        $t: (key) => key,
      },
      propsData: {
        page: 1,
        itemsCount: 30,
        itemsPerPage: 10,
      },
      localVue,
    });

    // Tested of die Komponente erstellt werden konnte
    expect(wrapper.vm).toBeTruthy();

    // Prüft ob die Komponente zwei Buttons enthält
    const buttons = wrapper.findAll('button');
    expect(buttons.length).toBe(2);

    // Klicke auf beide Buttons
    buttons.at(0).trigger('click');
    buttons.at(1).trigger('click');
    await wrapper.vm.$nextTick();

    // Jeder Button muss einmal das Page Changed event auslösen
    expect(wrapper.emitted().pageChanged.length).toBe(2);

    expect(wrapper.element).toMatchSnapshot();
  });
});
