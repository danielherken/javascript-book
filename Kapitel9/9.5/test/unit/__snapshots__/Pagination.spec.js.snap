// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination is emitting page changed event 1`] = `
<div
  class="row"
>
  <div
    class="col text-center"
  >
    <button
      class="btn btn-primary"
      type="button"
    >
      
      previous
    
    </button>
     
    <button
      class="btn btn-primary"
      type="button"
    >
      
      next
    
    </button>
  </div>
</div>
`;

exports[`Pagination is handling first page correctly 1`] = `
<div
  class="row"
>
  <div
    class="col text-center"
  >
    <button
      class="btn btn-primary disabled"
      disabled="disabled"
      type="button"
    >
      
      previous
    
    </button>
     
    <button
      class="btn btn-primary"
      type="button"
    >
      
      next
    
    </button>
  </div>
</div>
`;

exports[`Pagination is handling last page correctly 1`] = `
<div
  class="row"
>
  <div
    class="col text-center"
  >
    <button
      class="btn btn-primary"
      type="button"
    >
      
      previous
    
    </button>
     
    <button
      class="btn btn-primary disabled"
      disabled="disabled"
      type="button"
    >
      
      next
    
    </button>
  </div>
</div>
`;

exports[`Pagination is handling middle page correctly 1`] = `
<div
  class="row"
>
  <div
    class="col text-center"
  >
    <button
      class="btn btn-primary"
      type="button"
    >
      
      previous
    
    </button>
     
    <button
      class="btn btn-primary"
      type="button"
    >
      
      next
    
    </button>
  </div>
</div>
`;
